torchao/prototype/dtypes/uint2.py

```diff
@@ -0,0 +1,446 @@
+    __torch_dispatch__ doesn't guarantee the number of arguments you are
+    passed (e.g., defaulted arguments are not passed); but usually it is
+    convenient to pad out the arguments list with defaults.  This function
+    helps you do that.
+    Args:
+        args: the list of positional arguments passed to __torch_dispatch__
+        n: the number of arguments you are expecting to get
+        defaults_tail: default values for the arguments, starting from the
+            end of the list
+    Example:
+        >>> fill_defaults([1, 2, 3], 5, [3, 4, 5])
+        [1, 2, 3, 4, 5]
+        >>> fill_defaults([1, 2, 3], 5, [None, None, None])
+        [1, 2, 3, None, None]]
+    """
+    if n - len(defaults_tail) > len(args):
+        raise RuntimeError("not enough defaults to fill arguments")
+    r = list(args)
+    for i in range(len(args), n):
+        r.append(defaults_tail[i - n + len(defaults_tail)])
+    return r
+
+
+# qtensor_lib = Library("qtensors", "DEF")
+qtensor_lib.define(
+    "quantize_per_tensor_uint2(Tensor input, float scale, int zero_point) -> Tensor"
+)
+
+
+@impl(qtensor_lib, "quantize_per_tensor_uint2", "CompositeExplicitAutograd")
+def quantize_per_tensor_uint2(
+    input: torch.Tensor,
+    scale: float = 1.0,
+    zero_point: int = 1,
+) -> torch.Tensor:
+    inv_scale = 1.0 / scale
+    return pack_uint2(
+        torch.clamp(torch.round(input * inv_scale) + zero_point, 0, 2).to(torch.uint8)
+    )
+
+
+qtensor_lib.define(
+    "dequantize_per_tensor_uint2(Tensor input, float scale, int zero_point) -> Tensor"
+)
+
+
+@impl(qtensor_lib, "dequantize_per_tensor_uint2", "CompositeExplicitAutograd")
+def dequantize_per_tensor_uint2(
+    input: torch.Tensor,
+    scale: float = 1.0,
+    zero_point: int = 1,
+) -> torch.Tensor:
+    input = unpack_uint2(input)
+    return (input.view(torch.uint8).to(torch.float32) - zero_point) * scale
+
+
+class UInt2Tensor(torch.Tensor):
+    @staticmethod
+    def __new__(cls, elem, **kwargs):
+        assert elem.dtype is torch.uint8
+        assert not kwargs.get("requires_grad", False)
+        kwargs["requires_grad"] = False
+
+        return torch.Tensor._make_wrapper_subclass(
+            cls, up_size(elem.shape), dtype=torch.uint2, **kwargs
+        )
+
+    def __init__(self, elem, **kwargs):
+        self.elem = elem
+
+    @classmethod
+    def from_unpacked(cls, unpacked):
+        return UInt2Tensor(pack_uint2(unpacked))
+
+    def tolist(self):
+        return self.to(torch.uint8).tolist()
+
+    def __tensor_flatten__(self):
+        return ["elem"], None
+
+    @staticmethod
+    def __tensor_unflatten__(flattened, meta, outer_size, outer_stride):
+        assert meta is None
+        elem = flattened["elem"]
+        return UInt2Tensor(elem)
+
+    def __hash__(self):
+        return hash(self.elem)
+
+    def __eq__(self, other):
+        return torch.equal(self.elem, other.elem)
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs=None):
```

Comment:
Not feedback to you guys but I kind of hate how this looks - @cpuhrsch @jerryzh168 is there a more readable design pattern we could use? Something like an abstract class perhaps? Or a Dispatcher class?

Reply:
Yes, I like the `implements` pattern from NF4 and AQT.

https://github.com/pytorch/ao/blob/12f44abe0b4268816be3c29e172c43b727d8016e/torchao/dtypes/nf4tensor.py#L46-L54

Then you can do stuff like

```
@implements([torch.ops.aten.to.dtype])
def to_dtype(func, *args, **kwargs):
    if not args[0][0].is_contiguous():
        assert args[0][0].t().is_contiguous()
        return torch.ops.aten.to.dtype(args[0][0].t(), args[0][1]).t()
    return args[0][0].get_original_weight().to(args[0][1])
```

People can then also use this wrapper out of tree like in our tutorial: https://github.com/pytorch/ao/blob/12f44abe0b4268816be3c29e172c43b727d8016e/tutorials/add_an_op.py#L24

Reply:
torch_dispatch then becomes fairly simple

https://github.com/pytorch/ao/blob/12f44abe0b4268816be3c29e172c43b727d8016e/torchao/dtypes/nf4tensor.py#L757-L782

Please note that raising an exception at the end is very important. Otherwise you'll just return `None`, which is valid Python, even though you might have meant to say "This isn't implemented".

Also, it's probably preferred to use `return NotImplemented` instead of raising an exception within `__torch_dispatch__`, because it'll get caught and handled by the PyTorch dispatcher.

Reply:
@cpuhrsch incorporated your feedback in [1fdeb91](https://github.com/pytorch/ao/pull/282/commits/1fdeb91038c24e06d1b9a5ccb8d2a49da88bdaf7). @msaroufim 
 Have a look if its ok.